File: src/main/java/uk/gov/justice/hmpps/prison/security/AuthenticationFacade.java
package uk.gov.justice.hmpps.prison.security;

import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.RegExUtils;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;
import uk.gov.justice.hmpps.prison.web.config.AuthAwareAuthenticationToken;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Component
@Slf4j
public class AuthenticationFacade {

    public Authentication getAuthentication() {
        return SecurityContextHolder.getContext().getAuthentication();
    }

    public String getCurrentUsername() {
        final String username;

        final var userPrincipal = getUserPrincipal();

        if (userPrincipal instanceof String) {
            username = (String) userPrincipal;
        } else if (userPrincipal instanceof UserDetails) {
            username = ((UserDetails) userPrincipal).getUsername();
        } else if (userPrincipal instanceof Map userPrincipalMap) {
            username = (String) userPrincipalMap.get("username");
        } else {
            username = null;
        }

        return username;
    }

    public Collection<? extends GrantedAuthority> getCurrentRoles() {
        return getAuthentication() != null ? getAuthentication().getAuthorities() : null;
    }

    public boolean isClientOnly() {
        final var authentication = getAuthentication();
        if (!(authentication instanceof AuthAwareAuthenticationToken)) return false;

        return ((AuthAwareAuthenticationToken) authentication).isClientOnly();
    }

    public String getClientId() {
        final var authentication = getAuthentication();
        if (!(authentication instanceof AuthAwareAuthenticationToken)) return null;

        return (String)((AuthAwareAuthenticationToken) authentication).getClientId();
    }

    public String getGrantType() {
        final var authentication = getAuthentication();
        if (!(authentication instanceof AuthAwareAuthenticationToken)) return null;

        return (String)((AuthAwareAuthenticationToken) authentication).getGrantType();
    }

    public AuthSource getAuthenticationSource() {
        final var auth = getAuthentication();
        return Optional.ofNullable(auth).
                filter(AuthAwareAuthenticationToken.class::isInstance).
                map(AuthAwareAuthenticationToken.class::cast).
                filter(AbstractAuthenticationToken::isAuthenticated).
                map(AuthAwareAuthenticationToken::getAuthSource).
                orElse(AuthSource.NONE);
    }

    public static boolean hasRoles(final String... allowedRoles) {
        return hasMatchingRole(getRoles(allowedRoles), SecurityContextHolder.getContext().getAuthentication());
    }

    public boolean isOverrideRole(final String... overrideRoles) {
        return hasMatchingRole(getRoles(overrideRoles), getAuthentication());
    }

    private static boolean hasMatchingRole(final List<String> roles, final Authentication authentication) {
        return authentication != null &&
                authentication.getAuthorities().stream()
                        .anyMatch(a -> roles.contains(RegExUtils.replaceFirst(a.getAuthority(), "ROLE_", "")));
    }

    private Object getUserPrincipal() {
        final var auth = getAuthentication();
        return auth != null ? auth.getPrincipal() : null;
    }

    private static List<String> getRoles(String... roles){
        return Arrays.stream(roles).map(r -> RegExUtils.replaceFirst(r, "ROLE_", "")).toList();
    }
}

File: src/main/java/uk/gov/justice/hmpps/prison/security/AuthenticationFacade.kt
package uk.gov.justice.hmpps.prison.security

import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import uk.gov.justice.hmpps.kotlin.auth.AuthAwareAuthenticationToken
import uk.gov.justice.hmpps.kotlin.auth.AuthSource

@Component
class AuthenticationFacade {
  /**
   * This will return null if the token hasn't come from auth.  This is fine for application code, but tests need to
   * then use @WithMockAuthUser rather than using a TestingAuthenticationToken or @WithMockUser annotation.
   */
  val authentication: AuthAwareAuthenticationToken?
    get() = SecurityContextHolder.getContext().authentication as? AuthAwareAuthenticationToken

  /**
   * This is nullable since this can be called from an unprotected endpoint, but in the majority of cases it should
   * be not null.  This gets the current username from the authentication, falling back to the clientId if there
   * isn't a username passed in.
   */
  val currentPrincipal: String?
    get() = authentication?.principal

  val currentRoles: Collection<GrantedAuthority?>?
    get() = authentication?.authorities

  val isClientOnly: Boolean
    get() = authentication?.isSystemClientCredentials() ?: false

  val clientId: String?
    get() = authentication?.clientId

  /**
   * We are gradually moving away from user tokens and instead using client credentials more often.  This will be NONE
   * if client credentials are used, even if a NOMIS username is passed in.  This means that this method isn't an
   * adequate test to see if the user is a NOMIS user. The only test at present is to look the user up in the database
   * too see if they exist.
   */
  val authenticationSource: AuthSource
    get() = authentication?.authSource ?: AuthSource.NONE

  fun isOverrideRole(vararg overrideRoles: String): Boolean =
    hasMatchingRole(getRoles(*overrideRoles), authentication)

  companion object {
    fun hasRoles(vararg allowedRoles: String): Boolean =
      hasMatchingRole(getRoles(*allowedRoles), SecurityContextHolder.getContext().authentication)

    private fun hasMatchingRole(roles: List<String>, authentication: Authentication?): Boolean =
      authentication?.authorities?.any { roles.contains(it?.authority?.replaceFirst("ROLE_", "")) }
        ?: false

    private fun getRoles(vararg roles: String): List<String> = roles.map { it.replaceFirst("ROLE_", "") }
  }
}
